tests/test_web_tools.py

```
@@ -0,0 +1,63 @@
+        request = httpx.Request(method, url, headers=headers, extensions=extensions)
+        if headers["Host"] == "public.example" and request.url.path == "/start":
+            return httpx.Response(302, headers={"location": "http://127.0.0.1/private"}, request=request)
+        return httpx.Response(200, request=request)
+
+
+def test_redirect_target_is_validated_before_request(monkeypatch) -> None:
+    requested: list[str] = []
+
+    original_resolver = web_tools.resolve_public_http_url
+
+    def resolve(url: str):
```

Ruff (0.16.1)

[warning] Missing return type annotation for private function `resolve`

(ANN202)